DllImportGenerator/DllImportGenerator.UnitTests/ManualTypeMarshallingAnalyzerTests.cs

```diff
@@ -466,6 +466,102 @@ public Native(S s) : this()
                 VerifyCS.Diagnostic(NativeTypeMustBePointerSizedRule).WithSpan(24, 5, 24, 42).WithArguments("int", "S"));
         }
 
+        [Fact]
+        public async Task TypeWithGetPinnableReferencePointerReturnType_DoesNotReportDiagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+
+[NativeMarshalling(typeof(Native))]
+class S
+{
+    public byte c;
+
+    public ref byte GetPinnableReference() => ref c;
+}
+
+unsafe struct Native
+{
+    private IntPtr value;
+
+    public Native(S s) : this()
+    {
+        value = IntPtr.Zero;
+    }
+
+    public S ToManaged() => new S();
+
+    public int* Value { get => null; set {} }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(source);
+        }
+    
+        [Fact]
+        public async Task TypeWithGetPinnableReferenceByRefReturnType_ReportsDiagnostic()
+        {
+            string source = @"
+using System;
+using System.Runtime.InteropServices;
+
+[NativeMarshalling(typeof(Native))]
+class S
+{
+    public byte c;
+
+    public ref byte GetPinnableReference() => ref c;
+}
+
+unsafe struct Native
+{
+    private S value;
+
+    public Native(S s) : this()
+    {
+        value = s;
+    }
+
+    public ref byte Value { get => ref value.GetPinnableReference(); }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(source,
+                VerifyCS.Diagnostic(NativeTypeMustBePointerSizedRule).WithSpan(22, 5, 22, 71).WithArguments("ref byte", "S"),
```

Comment:
The analyzer testing APIs support [marking up the source text](https://github.com/dotnet/roslyn-sdk/blob/506767dc02d71f40b179f3e6b1e97b5d0f975b00/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/TestFileMarkupParser.cs) with expected diagnostics. For the tests in roslyn-analyzers, the general guideline is to use the markup if possible.

If doing argument validation, you still need to create/pass in the expected diagnostics, but I do find it simpler to at least not have to deal with the span line/column numbers. `GeneratedDllImportAnalyzerTests` has some tests that use the markup just for the location and combine it with the argument validation.

Not advocating for switching them all in this change - just for future reference.

Reply:
I'll do this in a follow up PR